tools/snapshot/snapshot_builder.cc

```diff
@@ -63,42 +69,60 @@ const std::vector<size_t>* NodeMainInstance::GetIsolateDataIndexes() {
 std::string SnapshotBuilder::Generate(
     const std::vector<std::string> args,
     const std::vector<std::string> exec_args) {
-  // TODO(joyeecheung): collect external references and set it in
-  // params.external_references.
-  std::vector<intptr_t> external_references = {
-      reinterpret_cast<intptr_t>(nullptr)};
   Isolate* isolate = Isolate::Allocate();
   per_process::v8_platform.Platform()->RegisterIsolate(isolate,
                                                        uv_default_loop());
   std::unique_ptr<NodeMainInstance> main_instance;
   std::string result;
+  int exit_code = 0;
+
+  HeapExternalReferences heap_external_references =
+      AllocateExternalRerefences();
+
+  std::vector<intptr_t> external_references = ExternalReferences::get_list();
+  external_references.insert(
+      external_references.end(),
+      heap_external_references.references.begin(),
+      heap_external_references.references.end());
+  external_references.push_back(ExternalReferences::kEnd);
 
   {
-    std::vector<size_t> isolate_data_indexes;
     SnapshotCreator creator(isolate, external_references.data());
+    SnapshotCreateData snapshot_data(&creator);
+    DeleteFnPtr<Environment, FreeEnvironment> env;
     {
       main_instance =
           NodeMainInstance::Create(isolate,
                                    uv_default_loop(),
                                    per_process::v8_platform.Platform(),
                                    args,
-                                   exec_args);
+                                   exec_args,
+                                   std::move(
+                                       heap_external_references.allocations));
       HandleScope scope(isolate);
       creator.SetDefaultContext(Context::New(isolate));
-      isolate_data_indexes = main_instance->isolate_data()->Serialize(&creator);
 
-      size_t index = creator.AddContext(NewContext(isolate));
+      env = main_instance->CreateMainEnvironment(&exit_code);
+      env->isolate_data()->Serialize(&snapshot_data);
+      env->Serialize(&snapshot_data);
+      snapshot_data.PrintErrorsAndAbortIfAny();
+      size_t index = creator.AddContext(env->context(),
+          { BaseObject::SerializeInternalFields, &snapshot_data });
       CHECK_EQ(index, NodeMainInstance::kNodeContextIndex);
+      CHECK_EQ(exit_code, 0);
     }
 
     // Must be out of HandleScope
     StartupData blob =
         creator.CreateBlob(SnapshotCreator::FunctionCodeHandling::kClear);
-    CHECK(blob.CanBeRehashed());
+    snapshot_data.PrintErrorsAndAbortIfAny();
+    // XXX(addaleax) disabling this check is not good
+    // CHECK(blob.CanBeRehashed());
```

Comment:
I have a WIP here https://chromium-review.googlesource.com/c/v8/v8/+/2143983 for this, BTW (verified with DOMException maps here https://github.com/joyeecheung/node/commits/test-hash)

Reply:
Cool! I’ll make sure to try that out as soon as I get to it.

Reply:
I don’t know if this is helpful or something that you already know, but for e.g. `builtInObjects` from `lib/internal/util/inspect.js`, this restore the contents of the Set, but `.has('Array')` returns false after deserialization even though it is listed in the Set

Reply:
hm, I can't reproduce this with v8 directly, will try to get a smaller reproduction from this branch

Reply:
Not yet gotten to the bottom of this but this seems to be related to the fact that we don't actually use the native Set/Maps in core, but the primordials ones. Somehow adding a line of `NativeModule.dummy = new Set([1]);` in the bootstrap script would do the trick for Sets.

Reply:
Okay, let me know if this is something that you could use any help with. :+1: In any case, I’ve included your WIP patch here to get an overview over what works and what doesn’t.

Reply:
I did some additional digging and it seems

1. This does not reproduce if only loaders.js is in the snapshot
2. This does not reproduce if loaders.js is in the snapshot while `nativeModuleRequire('internal/util/inspect')` is also performed early there
3. It goes away with adding `NativeModule.internalBindingWhitelist = internalBindingWhitelist;` in the loaders.js (while including node.js in the snapshot as well)
4. Something done below the point of `process.config = JSONParse(internalBinding('native_module').config);` in node.js caused this - I am not able to go further at the moment because moving the rest into loaders.js result in uninitialized performance_states buffers. Not sure what caused this yet.

Reply:
Hm, it probably has to be an issue in the V8 snapshotting implementation, though? I can also try to dig a bit deeper later…

Reply:
So far the smallest repro I have is

1. Change to snapshot loaders only
2. Add this to the loaders

```js
const EventEmitter = nativeModuleRequire('events');
const origProcProto = primordials.ObjectGetPrototypeOf(process);
primordials.ObjectSetPrototypeOf(origProcProto, EventEmitter.prototype);
const timers = nativeModuleRequire('timers');
```

This seems somewhat random, my guess is this could be GC-dependent, or some map transition needs to be taken care of. I could try doing in-place rehashing and see if this goes away, but no matter each approach is taken a smaller repro would be useful for testing against this.

Reply:
Um, I think I've found one of the Sets that wasn't properly rehashed during deserialization, so something in the rehashing logic is probably wrong. Still need to debug a bit deeper to create a repro for it, though.